BluetoothLeChat/app/src/main/java/com/example/bluetoothlechat/bluetooth/ChatServer.kt

```diff
@@ -0,0 +1,287 @@
+            gatt?.let {
+                val success = it.writeCharacteristic(messageCharacteristic)
+                Log.d(TAG, "onServicesDiscovered: message send: $success")
+                if (success) {
+                    _messages.value = Message.LocalMessage(message)
+                }
+            } ?: run {
+                Log.d(TAG, "sendMessage: no gatt connection to send a message with")
+            }
+        }
+        return false
+    }
+
+    /**
+     * Function to setup a local GATT server.
+     * This requires setting up the available services and characteristics that other devices
+     * can read and modify.
+     */
+    private fun setupGattServer(app: Application) {
+        gattServerCallback = GattServerCallback()
+
+        gattServer = bluetoothManager.openGattServer(
+            app,
+            gattServerCallback
+        ).apply {
+            addService(setupGattService())
+        }
+    }
+
+    /**
+     * Function to create the GATT Server with the required characteristics and descriptors
+     */
+    private fun setupGattService(): BluetoothGattService {
+        // Setup gatt service
+        val service = BluetoothGattService(SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)
+        // need to ensure that the property is writable and has the write permission
+        val messageCharacteristic = BluetoothGattCharacteristic(
+            MESSAGE_UUID,
+            BluetoothGattCharacteristic.PROPERTY_WRITE,
+            BluetoothGattCharacteristic.PERMISSION_WRITE
+        )
+        service.addCharacteristic(messageCharacteristic)
+        val confirmCharacteristic = BluetoothGattCharacteristic(
+            CONFIRM_UUID,
+            BluetoothGattCharacteristic.PROPERTY_WRITE,
+            BluetoothGattCharacteristic.PERMISSION_WRITE
+        )
+        service.addCharacteristic(confirmCharacteristic)
+
+        return service
+    }
+
+    // region setup advertising
+    /**
+     * Start advertising this device so other BLE devices can see it and connect
+     */
+    private fun startAdvertisement() {
+        advertiser = adapter.bluetoothLeAdvertiser
+        Log.d(TAG, "startAdvertisement: with advertiser $advertiser")
+
+        if (advertiseCallback == null) {
+            advertiseCallback = DeviceAdvertiseCallback()
+
+            advertiser?.startAdvertising(advertiseSettings, advertiseData, advertiseCallback)
+        }
+    }
+
+    /**
+     * Stops BLE Advertising.
+     */
+    private fun stopAdvertising() {
+        Log.d(TAG, "Stopping Advertising with advertiser $advertiser")
+        advertiser?.stopAdvertising(advertiseCallback)
+        advertiseCallback = null
+    }
+
+    /**
+     * Returns an AdvertiseData object which includes the Service UUID and Device Name.
+     */
+    private fun buildAdvertiseData(): AdvertiseData {
+        /**
+         * Note: There is a strict limit of 31 Bytes on packets sent over BLE Advertisements.
```

Comment:
Do we have a link to where the 31 bytes specification is made? Asking because I was never sure what this limit was. Also, is it worth mentioning `requestMtu()` could allow for more bytes to be sent?

Reply:
The 31 bytes spec comes from the bluetooth specification. I'm trying to dig up a good link for it but it's surprisingly hard to find a decent source for the information.

Reply:
And just as I type that I find one. Section 2.3.1.1 of this document (on page 42 of the doc) specifies that the advertising data can be up to 31 bytes:

https://inst.eecs.berkeley.edu/~ee290c/sp18/note/BLE_Vol6.pdf

Reply:
Let me know if this link works and I'll add it in a comment. If not I can look for a better link

Reply:
that link is great!